File: leetcode/python/001002.find-common-characters.py
from collections import Counter


class Solution:
    # Time complexity: O(n)
    # Space complexity: O(n)
    def commonChars(self, words: list[str]) -> list[str]:
        counters = [Counter(w) for w in words]
        res = []
        for i in range(ord("a"), ord("z") + 1):
            ch = chr(i)
            count = float("inf")
            for counter in counters:
                count = min(count, counter[ch])
            if count > 0:
                for _ in range(int(count)):
                    res.append(ch)
        return res

    # Time complexity: O(n)
    # Space complexity: O(n)
    def commonChars1(self, words: list[str]) -> list[str]:
        counters = [Counter(w) for w in words]
        res = []
        for i in range(ord("a"), ord("z") + 1):
            ch = chr(i)
            count = min([counter[ch] for counter in counters])
            if count > 0:
                res.extend([ch] * count)
        return res

    # Time complexity: O(n)
    # Space complexity: O(1) -> max 26
    def commonChars2(self, words: list[str]) -> list[str]:
        counter = Counter(words[0])

        for word in words:
            cur_counter = Counter(word)
            for c in counter:
                counter[c] = min(counter[c], cur_counter[c])

        res = []
        for c in counter:
            res.extend([c] * counter[c])

        return res

File: leetcode/javascript/001002.find-common-characters.js
/**
 * @param {string[]} words
 * @return {string[]}
 * Time complexity: O(n)
 * Space complexity: O(1) -> max 26
 */
var commonChars = function (words) {
  if (words.length === 1) {
    return words[0].split("");
  }

  const countMap = getCountMap(words[0]);
  for (let i = 1; i < words.length; i++) {
    const newCountMap = getCountMap(words[i]);

    for (const [k, v] of countMap.entries()) {
      countMap.set(k, Math.min(v, newCountMap.get(k) ?? 0));
    }
  }

  const res = [];

  for (const [k, v] of countMap.entries()) {
    for (let i = 0; i < v; i++) {
      res.push(k);
    }
  }

  return res;
};

function getCountMap(str) {
  const countMap = new Map();

  for (const ch of str) {
    countMap.set(ch, (countMap.get(ch) ?? 0) + 1);
  }

  return countMap;
}
